fv3core/stencils/d2a2c_vect.py

```diff
@@ -41,26 +41,66 @@ def lagrange_interpolation_x_p1(qy: sd, qout: sd):
         qout = lagrange_x_func_p1(qy)
 
 
-@gtscript.function
-def avg_x(u):
-    return 0.5 * (u + u[0, 1, 0])
-
-
-@gtscript.function
-def avg_y(v):
-    return 0.5 * (v + v[1, 0, 0])
-
-
 @gtstencil()
 def avg_box(u: sd, v: sd, utmp: sd, vtmp: sd):
     with computation(PARALLEL), interval(...):
-        utmp = avg_x(u)
-        vtmp = avg_y(v)
+        utmp = 0.5 * (u + u[0, 1, 0])
+        vtmp = 0.5 * (v + v[1, 0, 0])
 
 
 @gtscript.function
-def contravariant(u, v, cosa, rsin):
-    return (u - v * cosa) * rsin
+def contravariant(v1, v2, cosa, rsin2):
+    """
+    Retrieve the contravariant component of the wind from its covariant
+    component and the covariant component in the "other" (x/y) direction.
+
+    For an orthogonal grid, cosa would be 0 and rsin2 would be 1, meaning
+    the contravariant component is equal to the covariant component.
+    However, the gnomonic cubed sphere grid is not orthogonal.
+
+    Args:
+        v1: covariant component of the wind for which we want to get the
+            contravariant component
+        v2: covariant component of the wind for the other direction,
+            i.e. y if v1 is in x, x if v1 is in y
+        cosa: cosine of the angle between the local x-direction and y-direction.
+        rsin2: 1 / sin(alpha), where alpha is the angle between the local
+            x-direction and y-direction
+
+    Returns:
+        v1_contravariant: contravariant component of v1
+    """
+    # From technical docs on FV3 cubed sphere grid:
```

Comment:
is there a web address we could provide here?

Reply:
Not presently. The existing docs don't have this section, which is in Lucas's ongoing draft.